client-v2/src/main/java/com/clickhouse/client/api/Client.java

```diff
@@ -1166,25 +1199,27 @@ public CompletableFuture<InsertResponse> insert(String tableName,
                         metrics.operationComplete();
                         metrics.setQueryId(queryId);
                         return new InsertResponse(metrics);
-                    } catch (NoHttpResponseException e) {
-                        if (i < maxRetries) {
-                            try {
-                                data.reset();
-                            } catch (IOException ioe) {
-                                throw new ClientException("Failed to get response", e);
-                            }
-                            LOG.warn("Failed to get response. Retrying.", e);
+                    } catch ( NoHttpResponseException | ConnectionRequestTimeoutException | ConnectTimeoutException e) {
```

Comment:
Is it a duplicated code? (i see in other inserts as well)

Reply:
yes. I left it as duplicate because handling it in dedicated method would not be much shorter. I will review this code later if it can be extracted to a single method.